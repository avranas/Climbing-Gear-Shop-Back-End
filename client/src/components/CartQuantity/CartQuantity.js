import { useEffect, useState } from "react";
import "../Quantity/Quantity.css";

//For use in the Cart page
//Props are a handleSelection function, cartItem, and a default value
const CartQuantity = (props) => {
  const [options, setOptions] = useState([]);
  const [selection, setSelection] = useState(props.defaultValue);

  useEffect(() => {
    const amountInStock =
      props.cartItem.product.productOptions[0].amountInStock;
    //Cap at amountInStock or 99
    const newOptions = [];
    /*
      If another user buys this item which makes this user's cart quantity >
      amountInStock, there won't be an option for defaultValue, so we must set
      the maximum amount of options to whichever value is greatest,
      defaultValue or amountInStock
    */
    let maxOptions = 0;
    if (props.defaultValue > amountInStock) {
      maxOptions = props.defaultValue;
    } else {
      maxOptions = amountInStock;
    }
    for (let i = 1; i <= maxOptions && i <= 30; i++) {
      newOptions.push(i);
    }
    setOptions(newOptions);
  }, [props.cartItem, props.defaultValue]);

  const handleChange = (e) => {
    setSelection(e.target.value);
    props.handleSelection(e, props.cartItem);
  };
  return (
    <div>
      <p className="option-label">Quantity</p>
      <select
        className="form-select form-select-sm quantity-select"
        onChange={handleChange}
        name="quantity"
        id="quantity"
        value={selection}
      >
        {options.map((i, key) => {
          return (
            <option key={key} value={i}>
              {i}
            </option>
          );
        })}
      </select>
    </div>
  );
};

export default CartQuantity;
