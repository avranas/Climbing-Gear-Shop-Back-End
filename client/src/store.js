import { configureStore } from "@reduxjs/toolkit";
import userReducer from './slices/userSlice';
import productListSlice from './slices/productListSlice';
import notificationSlice from './slices/notificationSlice';
import productSlice from "./slices/productSlice";
import cartSlice from "./slices/cartSlice";
import newestCartItemSlice from "./slices/newestCartItemSlice";

export default configureStore({
  reducer: {
    user: userReducer,
    productList: productListSlice,
    product: productSlice,
    notifications: notificationSlice,
    cart: cartSlice,
    newestCartItem: newestCartItemSlice
  },
});