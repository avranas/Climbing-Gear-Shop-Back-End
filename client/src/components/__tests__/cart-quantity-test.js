import { cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import store from "../../store";
import CartQuantity from "../CartQuantity/CartQuantity";

afterEach(() => {
  cleanup();
});

const cartQuantity = (
  <Provider store={store}>
    <BrowserRouter>
      <CartQuantity />
    </BrowserRouter>
  </Provider>
);

test("CartQuantity matches snapshot", async () => {
  const tree = renderer.create(cartQuantity).toJSON();
  expect(tree).toMatchSnapshot();
});
