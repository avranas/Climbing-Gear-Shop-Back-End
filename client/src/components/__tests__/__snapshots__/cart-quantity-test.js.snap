// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartQuantity matches snapshot 1`] = `
<div>
  <p
    className="option-label"
  >
    Quantity
  </p>
  <select
    className="form-select form-select-sm quantity-select"
    id="quantity"
    name="quantity"
    onChange={[Function]}
  />
</div>
`;
